jdaviz/configs/cubeviz/plugins/parsers.py

```diff
@@ -88,6 +89,26 @@ def parse_data(app, file_obj, data_type=None, data_label=None):
         raise NotImplementedError(f'Unsupported data format: {file_obj}')
 
 
+def _return_spectrum_with_correct_units(flux, wcs, metadata, hdulist):
+    # Upstream issue of WCS not using the correct units for data must
+    # be fixed here.
+    # Issue: https://github.com/astropy/astropy/issues/3658
+    # Check for the correct CUNIT3 value in PRIMARY and then SCI, since
+    # it can be in either.
+    sc = Spectrum1D(flux=flux, wcs=wcs, meta=metadata)
+    if ('CUNIT3' in hdulist['PRIMARY'].header and
```

Comment:
This is two copies of almost identical code. Could we take out the logic to decide which extension it is in, and then store that decision, and use it only once? Something like:

```python
if this:
    ext = 'SCI'
else:
    ext = 'PRIMARY'
sc = uc.process_unit_conversion(
    self=uc, spectrum=sc, new_flux=None, new_spectral_axis=u.Unit(hdulist[ext].header['CUNIT3']))
```

Reply:
p.s. I don't think you have to explicitly pass in `self=uc`.

Reply:
It hasn't been instantiated so it fails if I do not include that line.

Reply:
Why not use the instantiated one then?

https://github.com/spacetelescope/jdaviz/blob/eac981fc56ea9e4a243d1def0ee89e329e9055ad/jdaviz/app.py#L1633

Reply:
That code block wouldn't work. It would have to be
```python
if this:
    ext = 'SCI'
else:
    ext = 'PRIMARY'
if u.Unit(hdulist[ext].header['CUNIT3']) != sc.spectral_axis.unit:
    sc = uc.process_unit_conversion(
        self=uc, spectrum=sc, new_flux=None, new_spectral_axis=u.Unit(hdulist[ext].header['CUNIT3']))
return sc
```

Would you prefer I change it to that instead?

Reply:
What do you think? I do like the less repetition but it is your code, so I'll let you make the call. 🐱

Reply:
I think I found a good way to streamline this section, please take a look if you have time @pllim .

Reply:
Yes, looks better. Thanks!